cmd/sourced/cmd/web.go

```diff
@@ -88,6 +102,10 @@ Once source{d} is fully initialized, the UI will be available, by default at:
 
 	select {
 	case err := <-ch:
+		if shouldFailFast(err) {
```

Comment:
This is unnecessary.

Reply:
If removed, those errors will be wrapped by a new one, what's wrong in this case

Reply:
Oh, you did this just to avoid wrapping, sorry. What I meant is that there's no need to fail fast here as it's already the last statement of the function.

Reply:
yeah. it's a bit confusing

Reply:
I'll be happy if you propose a better name for this function that filters our errors
```go
workdir.ErrMalformed
dir.ErrNotExist
dir.ErrNotDirectory
```

Considering it again, it might be better: `isNamedError`? wdyt @se7entyse7en?

The underlying problem, which is not trivial imo, is that `compose.RunWithIO` can fail in many ways, most of them unknown (or at least for me). And when running `web` command, we consider as "expected" all those unknown errors, and we just retry...
But when the error is "known" (from the list above), there is no sense to retry, and we can fail fast with a proper error message.

And I think that there is also another underlying problem: since we're bubbling errors in different places, without distinguishing them, then providing meaningful errors is a difficult task, and then we find these kinds of weird error messages from time to time.

Reply:
I'd split the `openUI() error` function into:
- `waitForContainer() error` that has the for loop as you have modified with the `shouldFailFast`, and it only returns `nil` or _known errors_,
- `openUI() error` that only returns `nil` or _error opening UI_.

Then `OpenUI` becomes something like:
```
// OpenUI opens the browser with the UI.
func OpenUI(timeout time.Duration) error {
    ch := make(chan error)
    go func() {
        err := waitForContainer()
        if err != nil {
            ch <- err
            return
        }

        err <- openUI()
        if err != nil {
            return errors.Wrap(err, "an error occurred while opening the UI")
        }
    }()

    ...

    select {
        case err := <-ch:
            return err
        case <-time.After(timeout):
            ...
    }
}
```

WDYT? With this splitting, it should be clearer what is a known error from a generic error occurred when trying to open the UI.

Reply:
you could eventually wrap also the first error `an error occurred while waiting for the container to start` or something along these lines.

Reply:
I think that refactors exceeds the goal of this PR.
But I pushed 058f88ffffd57c5a31a553f4460e0bfbb277092e that might address your requirement about clarifying different errors that can be returned by `openUI`, and that will also avoid [your initial &mdash;and legit&mdash; concern](https://github.com/src-d/sourced-ce/pull/162#discussion_r307661948)  about
```go
if failFast(err) {
  return err
} else {
  return error.Wrap(err)
}
```

Reply:
would be that :point_up: enough?

Reply:
I'd prefer the refactor as it clearly separates the different types of errors, but it's not a big concern.

Reply:
I understand your point, and I also think that this function might be too big and with different responsibilities inside, but I do really prefer to avoid refactors when hotfixing. Thanks for being nitpicking (because it helps a lot), and also for being flexible in this case :bowing_man: